### DISCUSSION TITLE
How to call publish_message function from another file?

### DISCUSSION THREAD

**Discussion by xya22er** [pika/pika#1421] [Q&A]
Hi 
In [asynchronous_publisher_example.py](https://github.com/pika/pika/blob/1.3.x/examples/asynchronous_publisher_example.py) example
I have problem that I could not  interact with **publish_message()** function because  **self._connection.ioloop.start()** prevents the main code from interacting once it has been called. 
How can I call **publish_message()** function from another file to send a different message each time?

**Comment by lukebakken** [comment]
Hello and thanks for using RabbitMQ and Pika.

The best way for me to assist you is for you to create a git repository that I can clone to see what you have _tried_ to do. Feel free to use GitHub for this.

**Comment by lukebakken** [comment]
Please see this code:

https://github.com/lukebakken/rabbitmq-users-pika-fSrIb6df-yc

**Reply by xya22er** [reply]
Thanks for your replay and help 
I need something like that [github](https://github.com/xya22er/pika_asynchronous_publisher)

**Reply by lukebakken** [reply]
Please re-read my comment: [link](https://github.com/pika/pika/discussions/1421#discussioncomment-5220817). Unless you'd like to pay me to write code for you, you'll have to figure this out yourself.

I'm happy to review your code once you give something a try, and I will even contribute. What I won't do is write code _for_ you.

**Reply by xya22er** [reply]
Hi again,

I have solved my problem and now I am able to interact with **publish_message()** function. However, I am not sure if this is the optimal solution. Can you check the code [here](https://github.com/xya22er/pika_asynchronous_publisher) and give me your insights?

Thank you

**Reply by lukebakken** [reply]
Please check my most recent changes here:

https://github.com/lukebakken/rabbitmq-users-pika-fSrIb6df-yc/blob/main/connection.py

https://groups.google.com/g/rabbitmq-users/c/fSrIb6df-yc/m/48vj6NeIBgAJ

Notice that the general pattern is to schedule work via `connection.add_callback_threadsafe` then wait for a response, if you want to handle the response synchronously.
